### ISSUE TITLE
Test memory usage for grounding refactoring

### ISSUE THREAD

**Issue by hickst** [clulab/reach#490]
Perhaps this can be of use: https://github.com/jbellis/jamm

**Comment by hickst** [comment]
Tried 4 different methods of testing this: 2 failed utterly. 1 was very rough estimate base on GC and memory allocation.
JAMM lib succeeded but still needs much more testing as the implementations differ in the number and root data structures of KB objects. Initial, rough measure reported:
1,246,808,136 bytes for master
1,384,729,816 bytes for grefac

**Comment by hickst** [comment]
Retried the GC/memory alloc method and the results are **all over the map**, making it meaningless.

Refined the JAMM library method to report details. For the master branch:
```
          ContextCellLine = 39267464
         ContextCellLine2 = 51080
          ContextCellType = 3720000
             ContextOrgan = 19694832
           ContextSpecies = 892184
        ContextTissueType = 1089528
         StaticBioProcess = 129944
       StaticCellLocation = 1671424
      StaticCellLocation2 = 445248
           StaticChemical = 104883432
               StaticDrug = 203016
            StaticProtein = 1037612776
     StaticProteinComplex = 46808
     StaticProteinFamily0 = 145512
      StaticProteinFamily = 14968584
     StaticProteinFamily2 = 21807672
    ModelGendCellLocation = 1728
        ModelGendChemical = 38000
ModelGendProteinAndFamily = 144656
      ReversSpeciesLookup = 284600
                    TOTAL = 1247098488
               RIML + RSL = 1247092736
```
and for the refactored branch:
```
          ContextCellLine = 62978024
         ContextCellLine2 = 84792
          ContextCellType = 4903432
             ContextOrgan = 26424904
           ContextSpecies = 1200744
        ContextTissueType = 1574504
         StaticBioProcess = 169624
       StaticCellLocation = 2283512
      StaticCellLocation2 = 617376
           StaticChemical = 127984128
               StaticDrug = 339616
            StaticProtein = 1112633712
     StaticProteinComplex = 74528
     StaticProteinFamily0 = 233192
      StaticProteinFamily = 20498424
     StaticProteinFamily2 = 22432016
    ModelGendCellLocation = 2144
        ModelGendChemical = 61320
ModelGendProteinAndFamily = 236816
                    TOTAL = 1384732808
                     RIML = 1384729816
```

**Comment by hickst** [comment]
I'm surprised by how difficult it was to measure data structures in memory in the JVM. Assuming that we can trust the JAMM library numbers above (and I have no evidence one way or another), then the refactored grounding definitely increases memory consumption. This seems like another indicator that we should move to use a real (embedded) database (as per issue #482). However, I would still recommend we merge the refactored grounding code as it has several other advantageous features:

1. The grounding classes have been simplified and the code reduced from 34 files, 34 classes, and 2632 lines to 26 files, 30 classes and 2357 lines.
2. Separate, additional, "reverse lookup" KBs are no longer necessary as the capability to search for KB entries by NS/ID is now available for **all the KBs** by default.
3. The NS/ID search capability can be used to identify "synonym sets" for a given grounding NS/ID (i.e. sets of KB entry text strings which have the same grounding NS/ID).
4. Each KB is now configured from data structure arguments, which will make it possible to configure them externally from a configuration file (not yet implemented).

If there are no major objections, I'd like to merge the refactored code to the trunk.

**Comment by MihaiSurdeanu** [comment]
OK for me.

**Comment by hickst** [comment]
Closed by pull request #491

**Comment by hickst** [comment]
Note on usage of JAMM, since it must be run as a Java Agent:

`java -jar -javaagent:/Users/hickst/workspaceCLU/clulab/reach/main/lib/jamm-0.3.1.jar /opt/local/share/sbt/sbt-launch.jar 'run-main org.clulab.reach.ShowMemReachable'`
